### ISSUE TITLE
Upload Bug or lacking knowledge

### ISSUE THREAD

**Issue by johntom** [strapi/strapi#1753]
Node.js version: 9.11.1
npm version: 5.6.0
Strapi version: 3.0.0-alpha.13.1
Database: mongoDB v4.0.0
Operating system: win10
What is the current behavior?
Trying to port a trails server that uses muter to upload files to specific directories.
I'm testing wth 'koa-multer' and get the following

What is the current behavior?
Client: tried many variants but this seems like the best and putting headers in request caused more issues
    //     headers: {
    //  'Content-type': 'application/x-www-form-urlencoded; charset=UTF-8',
    //  'Accept': 'application/json',
    // or   'Content-type': 'multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW',
    //  },

   let formData = new FormData()
   //    console.log('fname ', image[0].name)
    formData.append('file', image[0]);
   let baseweb = 'http://localhost:1337/'
   var url = baseweb + `gym/upload`
    return fetch(url, {
      mode: 'cors',
      method: 'POST',
    //  body: JSON.stringify(formData) // does not work with koa
     body:formData
    }).then((res) => res.json())
      .then(data => console.log('data.message', data.message))
      .catch(error => console.log(error));
  },
```
Steps to reproduce the problem
After I upload a csv file I get the following.
```
 files:
   { file:
      File {
        domain: null,
        _events: {},
        _eventsCount: 0,
        _maxListeners: undefined,
        size: 1574,
        path: 'C:\\Users\\john\\AppData\\Local\\Temp\\upload_14d719cc6b8eaedf6d7f51f5ae303df5',
        name: 'Gym Door Report0722.csv',
        type: 'application/vnd.ms-excel',
        hash: null,
        lastModifiedDate: 2018-08-15T02:05:22.593Z,
        _writeStream: [WriteStream] } } }
[2018-08-15T02:05:22.610Z] debug POST /gym/upload (35 ms)
(node:14516) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'headers' of undefined
    at hasbody (D:\Frameworks\Strapi-Gatsby\api\node_modules\type-is\index.js:93:14)
    at typeofrequest (D:\Frameworks\Strapi-Gatsby\api\node_modules\type-is\index.js:127:8)
    at multerMiddleware (D:\Frameworks\Strapi-Gatsby\api\node_modules\multer\lib\make-middleware.js:18:10)
    at Promise (D:\Frameworks\Strapi-Gatsby\api\node_modules\koa-multer\index.js:38:9)
    at new Promise (<anonymous>)
    at D:\Frameworks\Strapi-Gatsby\api\node_modules\koa-multer\index.js:37:14
    at upload (D:\Frameworks\Strapi-Gatsby\api\api\gym\controllers\Gym.js:112:25)
    at dispatch (D:\Frameworks\Strapi-Gatsby\api\node_modules\koa-router\node_modules\koa-compose\index.js:44:32)
    at next (D:\Frameworks\Strapi-Gatsby\api\node_modules\koa-router\node_modules\koa-compose\index.js:45:18)
    at dispatch (D:\Frameworks\Strapi-Gatsby\api\node_modules\koa-compose\index.js:42:32)
    at policies.push (D:\Frameworks\Strapi-Gatsby\api\node_modules\strapi\lib\middlewares\router\utils\routerChecker.js:59:26)
    at dispatch (D:\Frameworks\Strapi-Gatsby\api\node_modules\koa-compose\index.js:42:32)
    at module.exports (D:\Frameworks\Strapi-Gatsby\api\plugins\users-permissions\config\policies\permissions.js:55:9)
    at <anonymous>
(node:14516) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:14516) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
What is the expected behavior?
upload 1 or more files to directory, don't always need a file link


Suggested solutions
Current Code
```
'use strict';

/**
 * Gym.js controller
 *
 * @description: A set of functions called "actions" for managing `Gym`.
 */
```
var multer = require('koa-multer');
const path = require('path')
const fs = require('fs.extra')

// const koaBody = require('koa-body')({ multipart: true });
// const bodyParser = require('koa-bodyparser');

const storage = multer.diskStorage({
  destination: function (req, file, cb) {

    let dirtest = 'c:/DailyImport/import'
    let fpath = dirtest + '/' + file.originalname
    console.log('path ', fpath);

    if (!fs.existsSync(dirtest)) {
      //this.log.info("make dir");
      fs.mkdirp(dirtest, function (err) {
        if (err) {
          // this.log.info(err);
        } else {
          // this.log.info('pow!')
          cb(null, dirtest)
        }
      })
    } else cb(null, dirtest)

  },
  filename: function (req, file, cb) {
    console.log(' file.originalname', file, file.originalname)
    cb(null, file.originalname)
  },

})

var upload = multer({
  storage: storage,
  fileFilter: function (req, file, cb) {
    console.log('upload file.mimetype', file.mimetype)

    if (file.mimetype === 'application/pdf' || file.mimetype !== 'application/xls' || file.mimetype !== 'text/csv') {
      cb(null, true);
    } else {
      req.fileValidationError = 'goes wrong on the mimetype';
      return cb(null, false, new Error('goes wrong on the mimetype'));
    }
  }
})

module.exports = {
  find: async (ctx) => {
    // return strapi.services.gym.fetchAll(ctx.query);
    const data = await strapi.services.gym.fetchAll(ctx.query);
    console.log('data ', data.length)
    return data // or  ctx.send(data);//or  // Send 200 `ok`
  },
  upload: async (ctx) => {
    console.log('upload files ')
    console.log('=======================')
    // let rr = path.resolve(__dirname, '..') // api dir
    let rr = path.resolve(__dirname, '../..') // 1 up  api dir
    let mess = ' files uploaded'
  
    let body =ctx.request.body 
    // of below same error
   //let body =JSON.stringify(ctx.request.body) //bodyParser(ctx.request.body)
    console.log('======================sbody=', body)
    upload.array('file')(body, ctx.response, err => {
      if (err) {
        console.log('err ', err)
        this.log.info(err)
      }
      console.log('req.files[i req.files]', ctx.request.files)
      // let mess = flen + ' files uploaded'
      return { 'message': mess }// res.json({ 'message': mess })
      ctx.send(mess)//'Hello World!');
    })
  },

// below for mongo aggregates written to collection

  aggregate: async (ctx) => {
     const data = await strapi.services.gym.fetchAllaggregate(ctx.query);
    console.log('Gym.js: aggregate data ', data.length)
    let obj;
    let objarray = []
    data.forEach(function (rec) {
      var userStr = JSON.stringify(rec);
      obj = {};
      JSON.parse(userStr, (key, value) => {
        if (typeof value === 'string') {
          obj[key] = value;
        }
        if (typeof value === 'number') {
          obj[key] = value;
          if (key === 'count') {
            obj.CYM = obj.year + '-' + obj.month;
            console.log('obj  value ', obj.tenant, obj.year, obj.month, obj.count);
            objarray.push(obj)
            obj = {};
          }
        }
      });
    })
       console.log('objarray 1 ==========', objarray.length)
    await strapi.services.gymaggregate.removeall()
    await strapi.services.gymaggregate.bulkadd(objarray)
    return objarray
  },


  findOne: async (ctx) => {
    if (!ctx.params._id.match(/^[0-9a-fA-F]{24}$/)) {
      return ctx.notFound();
    }

    return strapi.services.gym.fetch(ctx.params);
  },



  count: async (ctx) => {
    return strapi.services.gym.count(ctx.query);
  },


  create: async (ctx) => {
    return strapi.services.gym.add(ctx.request.body);
  },

  update: async (ctx, next) => {
    return strapi.services.gym.edit(ctx.params, ctx.request.body);
  },

 

  destroy: async (ctx, next) => {
    return strapi.services.gym.remove(ctx.params);
  }
};
```

**Comment by derrickmehaffy** [comment]
@johntom Can you please edit this issue and use code blocks or external sources like pastebin?

Your issue is quite hard to follow

**Comment by johntom** [comment]
Sorry, I did use code blocks but had some misplaced ``` I will make a public github repo and try to clarify. At the end of the day I'd like to use koa-multer for uploading as I need total flexibilty on the process.
Thanks,
John

**Comment by johntom** [comment]
Hi,
The follow repo https://github.com/johntom/strapiApp is now available
1) \strapi\api\api\gym\controllers\Gym.js has the upload method using
koa-multer
2) client uses fetch

let formData = new FormData()

formData.append('file', image[0]);
let baseweb = 'http://localhost:1337/'
var url = baseweb + gym/upload
return fetch(url, {
mode: 'cors',
method: 'POST',
// body: JSON.stringify(formData) // does not work with koa
body:formData
}).then((res) => res.json())
.then(data => console.log('data.message', data.message))
.catch(error => console.log(error));
}

3) error
files:
{ file:
File {
domain: null,
_events: {},
_eventsCount: 0,
_maxListeners: undefined,
size: 1574,
path:
'C:\Users\john\AppData\Local\Temp\upload_14d719cc6b8eaedf6d7f51f5ae303df5',
name: 'Gym Door Report0722.csv',
type: 'application/vnd.ms-excel',
hash: null,
lastModifiedDate: 2018-08-15T02:05:22.593Z,
_writeStream: [WriteStream] } } }
[2018-08-15T02:05:22.610Z] debug POST /gym/upload (35 ms)
(node:14516) UnhandledPromiseRejectionWarning: TypeError: Cannot read
property 'headers' of undefined
4) when if working happy to leave repo as an exmaple of multer and
aggragation
TIA

John

**Comment by lauriejim** [comment]
You can check the Strapi SDK JS we have on our repo.
